### PR TITLE
fix: Upgrade azserect client for AZURE

### PR DESCRIPTION
**Pull request by jiuker**
Upgrade azserect client
How to test:
```shell
export ClientID=UUID
export EndPoint=url
export Secret=secret
export TenantID=tid
export ManagedIdentityClientID=micd
go test github.com/minio/kes/internal/keystore/azure
```
Once the verification is successful, I remove the redundant code
fix https://github.com/minio/kes/issues/261

### CODE REVIEW

File: internal/keystore/azure/key-vault-error.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package azure
```
Comment (aead) [inline internal/keystore/azure/key-vault-error.go:1]: Missing license and copyright error.

File: internal/keystore/azure/key-vault-error.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"net/http"
+	"reflect"
+)
+
+// ResponseError is a wrapper around an error response from the Azure Key Vault service.
+type ResponseError struct {
```
Comment (aead) [inline internal/keystore/azure/key-vault-error.go:9]: Does this need to be exported or this only an internal error?

File: internal/keystore/azure/key-vault-error.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	StatusCode int
+
+	// RawResponse is the underlying HTTP response.
+	RawResponse *http.Response
+
+	errorResponse errorResponse
+}
+
+// TransportErrToResponseError converts a transport error to a ResponseError.
+func TransportErrToResponseError(terr error) (*ResponseError, bool) {
```
Comment (aead) [inline internal/keystore/azure/key-vault-error.go:23]: Same, does this need to be exported?
Also, what exactly is a transport error and what azure API error type are we handling here?

File: internal/keystore/azure/key-vault.go [hunk 4]
```diff
@@ -338,6 +369,33 @@ func (s *Store) Close() error { return nil }
 // ConnectWithCredentials tries to establish a connection to a Azure KeyVault
 // instance using Azure client credentials.
 func ConnectWithCredentials(_ context.Context, endpoint string, creds Credentials) (*Store, error) {
+	if os.Getenv("AZURE_CLIENT_API_VERSION") != "7.2" {
```
Comment (aead) [inline internal/keystore/azure/key-vault.go:372]: Do we have to pass them as env var? Can't we pass them directly as credentials?

File: internal/keystore/azure/key-vault.go [hunk 5]
```diff
@@ -358,6 +416,32 @@ func ConnectWithCredentials(_ context.Context, endpoint string, creds Credential
 // ConnectWithIdentity tries to establish a connection to a Azure KeyVault
 // instance using an Azure managed identity.
 func ConnectWithIdentity(_ context.Context, endpoint string, msi ManagedIdentity) (*Store, error) {
+	if os.Getenv("AZURE_CLIENT_API_VERSION") != "7.2" {
```
Comment (aead) [inline internal/keystore/azure/key-vault.go:419]: Why are we requiring API version '7.2' here. Would this break if the API version gets bumped to e.g. 7.3

**Review by aead — CHANGES_REQUESTED** [review]

File: internal/keystore/azure/key-vault.go (hunk 5 above)
Comment (jiuker) [reply internal/keystore/azure/key-vault.go]: This is just a redundant judgment to determine whether the behavior of two clients is consistent. If they are consistent, the rest code will be deleted. You need to make a judgment here, as I am not familiar with this @aead

File: internal/keystore/azure/client.go [hunk 6]
```diff
@@ -108,6 +131,54 @@ func (c *client) CreateSecret(ctx context.Context, name, value string) (status,
...
+				StatusCode: azResp.StatusCode,
+				ErrorCode:  azResp.ErrorCode,
+				Message:    azResp.errorResponse.Error.Message,
+			}, nil
+		}
+		if response.Attributes.Enabled != nil && !*response.Attributes.Enabled {
+			return "", status{
+				StatusCode: http.StatusUnprocessableEntity,
+				ErrorCode:  "ObjectIsDisabled",
+				Message:    fmt.Sprintf("The secret %q is current disabled and cannot be used", name),
```
Comment (shtripat) [inline internal/keystore/azure/client.go:154]: `%s/current/currently/g` or just remove the word `current`

**Comment by bh4t** [comment]
Hey @aead can you please review this when you get a chance?

File: internal/keystore/azure/key-vault-error.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2021 - MinIO, Inc. All rights reserved.
```
Comment (harshavardhana) [inline internal/keystore/azure/key-vault-error.go:1]:
```suggestion
// Copyright 2024 - MinIO, Inc. All rights reserved.
```

**Review by harshavardhana — COMMENTED** [review]
Only minor comment, rest of them looks fine. We need to get rid of the old code @jiuker

**Comment by vadmeste** [comment]
@jiuker did you get the chance to test this with Azure ? let me know if you need any help in Azure

**Comment by jiuker** [comment]
> @jiuker did you get the chance to test this with Azure ? let me know if you need any help in Azure

Yeah. Pre commit have test code. I have tested it. @vadmeste

**Comment by harshavardhana** [comment]
@aead we need to get this merged for a new kes release, so that we can move to a supported SDKs.  Can you PTAL ?

**Comment by bh4t** [comment]
@aead can you please review this?

**Review by harshavardhana — DISMISSED** [review]

**Comment by harshavardhana** [comment]
@aead ^^

**Comment by harshavardhana** [comment]
Checking back again on this @aead let me know if you could look at this PR.

**Comment by harshavardhana** [comment]
https://github.com/minio/kes/pull/459 supersedes this PR, closing this one.

**Pull request closed without merging**